File: lib/data/search_logic.dart
// Pure query-kind detection (no data). Buses + Stops are resolved live by
// DataStore.searchServices / DataStore.searchStops.
//
// Direct port of legacy/ios-native/Lyne/SearchLogic.swift — same regexes,
// same precedence.

import 'models.dart';

DetectedKind detectQueryKind(String raw) {
  final q = raw.trim();
  if (q.isEmpty) return const DetectedKind('empty', '');

  if (RegExp(r'^\d{6}$').hasMatch(q)) {
    return const DetectedKind('postal', 'Postal code');
  }
  if (RegExp(r'^\d{5}$').hasMatch(q)) {
    return const DetectedKind('stopcode', 'Stop code');
  }
  // Up to 2 leading letters covers real SG services: 88, 410W, NR1, CT8.
  if (RegExp(r'^[A-Za-z]{0,2}\d{1,3}[A-Za-z]?$').hasMatch(q)) {
    return const DetectedKind('bus', 'Bus service');
  }
  if (RegExp(r'^blk\s*\d', caseSensitive: false).hasMatch(q)) {
    return const DetectedKind('block', 'Block + street');
  }
  return const DetectedKind('text', 'Name or place');
}

File: legacy/ios-native/Lyne/SearchLogic.swift
// Pure query-kind detection (no data). Buses + Stops are resolved live by
// DataStore.searchServices / DataStore.searchStops.

import Foundation

func detectQueryKind(_ raw: String) -> DetectedKind {
    let q = raw.trimmingCharacters(in: .whitespaces)
    if q.isEmpty { return DetectedKind(kind: "empty", label: "") }
    if q.range(of: #"^\d{6}$"#, options: .regularExpression) != nil {
        return DetectedKind(kind: "postal", label: "Postal code")
    }
    if q.range(of: #"^\d{5}$"#, options: .regularExpression) != nil {
        return DetectedKind(kind: "stopcode", label: "Stop code")
    }
    // Up to 2 leading letters covers real SG services: 88, 410W, NR1, CT8.
    if q.range(of: #"^[A-Za-z]{0,2}\d{1,3}[A-Za-z]?$"#, options: .regularExpression) != nil {
        return DetectedKind(kind: "bus", label: "Bus service")
    }
    if q.range(of: #"^blk\s*\d"#, options: [.regularExpression, .caseInsensitive]) != nil {
        return DetectedKind(kind: "block", label: "Block + street")
    }
    return DetectedKind(kind: "text", label: "Name or place")
}
